src/crystal/system/unix/file_descriptor.cr

```diff
@@ -236,19 +236,41 @@ module Crystal::System::FileDescriptor
 
   @[AlwaysInline]
   def self.tcgetattr(fd)
-    termios = uninitialized LibC::Termios
-    ret = {% if flag?(:android) && !LibC.has_method?(:tcgetattr) %}
-            LibC.ioctl(fd, LibC::TCGETS, pointerof(termios))
-          {% else %}
-            LibC.tcgetattr(fd, pointerof(termios))
-          {% end %}
-    raise IO::Error.from_errno("tcgetattr") if ret != 0
-    termios
+    {% if flag?(:solaris) %}
+      termio = uninitialized LibC::Termio
+      ret = LibC.ioctl(fd, LibC::TCGETA, pointerof(termio))
+      raise IO::Error.from_errno("ioctl") if ret != 0
+      termio
+    {% else %}
+      termios = uninitialized LibC::Termios
+      ret = {% if flag?(:android) && !LibC.has_method?(:tcgetattr) %}
+              LibC.ioctl(fd, LibC::TCGETS, pointerof(termios))
+            {% else %}
+              LibC.tcgetattr(fd, pointerof(termios))
+            {% end %}
+      raise IO::Error.from_errno("tcgetattr") if ret != 0
+      termios
+    {% end %}
```

Comment:
**question:** I'm wondering if this couldn't be merged in with the existing code.
There are a total of three different implementations of this method. The form is identical for all of them, just the core method and its arguments changes.
The code seems like it could make use of `out` to simplify as well. So technically, we don't even need to specify whether the output argument to `ioctl` is typed `LibC::Termio` or `LibC::Termios`. The compiler can derive that automatically.

Reply:
Ah, yeah that makes sense. Still, we could incorporate solaris in the same way as the current two branches.